### ISSUE TITLE
Conditional constant definition triggers reportConstantRedefinition

### ISSUE THREAD

**Issue by niun** [microsoft/pyright#5265]
**Describe the bug**
When defining a constant (upper case variable name) in two mutually exclusive execution paths of an if-else condition or a try except statement, the constant is reported to be redefined.

**To Reproduce**
Create a file `pyproject.toml` with the following content:
``` toml
[tool.pyright]
reportConstantRedefinition = true
```

Create the example script:

_Example A:_
``` python
# example_a.py

try:
    import some_module
except ImportError:
    __SOME_MODULE_SUPPORTED = False
else:
    __SOME_MODULE_SUPPORTED = True
```

Execute pyright:
``` shell
$ python -m pyright example_a.py
example_a.py
  example_a.py:4:12 - error: Import "some_module" could not be resolved (reportMissingImports)
  example_a.py:6:5 - error: "__SOME_MODULE_SUPPORTED" is constant (because it is uppercase) and cannot be redefined (reportConstantRedefinition)
2 errors, 0 warnings, 0 informations
```
The second reported error here is the relevant one.

Another example would be the following one:

_Example B:_
``` python
# example_b.py

import os

if os.getenv('DEBUG'):
    DEBUG_NOTIFICATION = "Debugging is enabled!"
else:
    DEBUG_NOTIFICATION = ""
```

``` shell
$ python -m pyright example_b.py
example_b.py
  example_b.py:8:5 - error: "DEBUG_NOTIFICATION" is constant (because it is uppercase) and cannot be redefined (reportConstantRedefinition)
1 error, 0 warnings, 0 informations
```

Of course there is a workaround for example B where the error is not triggered:
``` python
DEBUG_NOTIFICATION = "Debugging is enabled!" if os.getenv('DEBUG') else ""
```

**Expected behavior**
I would expect no redefinition error to be reported in both examples, because the execution paths for the two different definitions of the same constant are mutually exclusive.

**VS Code extension or command-line**
The unexpected behaviour is reproducible with the command-line tool and the VS Code Pylance extension (both using pyright version 1.1.313, did not try the pyright extension).

**Comment by erictraut** [comment]
This behavior is by design. It's not a bug. A constant (by definition) can have only one statically-determined value. If it can have different values at runtime based on dynamic behaviors, then it's a variable, not a constant.

I presume that you've enabled the strict-mode `reportConstantRedefinition` diagnostic check. By default, this check is not enabled in pyright. If you don't want pyright to detect redefinitions of constants, you can disable this check in your project.

**Comment by bukzor-sentryio** [comment]
Please reconsider? Your definition of constant above matches the idea of `Literal` but constants need not be literals.

> A constant (by definition) can have only one statically-determined value. If it can have different values at runtime based on dynamic behaviors, then it's a variable, not a constant.

For example, this is a perfectly valid constant, and `pyright` agrees:

```python
import os

DEBUG = bool(os.getenv('DEBUG', ''))
```

Consider the below example. `BAD` is flagged as "redefined" while `GOOD` is not, even though they both have type `Literal[1] | Literal[2]`. (In fact I wouldn't be surprised if pyright has done sufficient analysis to know that they are equal by construction.)

```python
from random import random

if random() > 0.5:
    BAD = const = 1
else:
    BAD = const = 2

GOOD = const
```

In other words type-broadening due to an `if` shouldn't count as reassignment, when each branch only has one assignment.  I quite like this check in principle but in its current form I'm forced to disable it when writing strict-but-idiomatic python.

**Comment by marcospgp** [comment]
ran into this today, any chance this could be reconsidered as the poster above also requested?

**Comment by niun** [comment]
Please consider reopening this issue. In Python I don't think the previously given definition of a constant is helpful as there isn't anything statically determined in Python? I think a python-tailored definition for a constant would be something like: "A constant is a variable whose value is determined and assigned once and is never changed afterwards."

In the original examples, the global constants are created once with a value which is never changed. There are two places in which the variable is declared. But as the places are in mutually exclusive paths where only one is ever executed, the evaluation and assignment of the value is done only once.

I earn my money with embedded C development. In C it is common to use the preprocessor like this to be able to build two different application versions:

    #ifdef HARDWARE_VERSION == 1
    const int output_pin_address = 0xDEAD
    #else
    const int output_pin_address = 0xBEAF
    #endif

Since Python has no compiling, no preprocessor and only runtime assignment, there is (luckily) no preprocessor syntax to do this, but it would be nice if I could write in a similar unobfuscated way with clean python syntax that there are different ways this application can be run based on environment conditions or arguments. And using a function or an intermediate variable to determine the value for those constants does feel like unnecessary obfuscation (and even less static) to me.

**Comment by kylebebak** [comment]
Just to expand on the examples above:

```py
from random import random

r = random()

if r > 0.5:
    BAD = 1
else:
    BAD = 2

GOOD = 1 if r > 0.5 else 2
```

`BAD` and `GOOD` are declared with effectively the same code, but with `reportConstantRedefinition = true` one of them is allowed and the other is not. I think this is inconsistent and must be surprising to most people who run into it

**Comment by RBerga06** [comment]
I'll add an use case to the list: platform-specific constants.
```python
# pyright: reportUnreachable=false
import sys

class MacFoo: pass
class WinFoo: pass
class LinFoo: pass


if sys.platform == "darwin":
    OS = "macOS"
    type SysFoo = MacFoo  # reportRedeclaration
    class SysBar: pass    # reportRedeclaration

elif sys.platform == "win32":
    OS = "Windows"        # reportConstantRedefinition
    type SysFoo = WinFoo  # reportRedeclaration
    class SysBar: pass    # reportRedeclaration

else:
    OS = "Linux"          # reportConstantRedefinition
    type SysFoo = LinFoo
    class SysBar: pass
```
([pyright playground](https://pyright-play.net/?strict=true&code=MQAgDgngTglg5gCwC4C4RQKZgPZSQVQDtMBDAYwRICMAbDAXgDMSaBnDAKBgFsc8RWEVhw5kaJVqxABZcgDFs2NGAnCxqkAHUYhBUvCrR4ySAAyOvcsMiYjAUIB0YcUka5uIevRAAiACYkUADuOj4oHCCRIADyAMqevtzkcT4RUUgQYBggsUJ6CbJk%2BSCgmHxIAEoYfhjqUCRIMNiEaZHqJrmsAEKBViZRpVi4ldW14vWNzSIYNLb2rE4ublAeXr4hhADMAExhrTHx3j7ahH7YQaw%2BUdcl6EN4AMLNrEgkhCM1jDowky3XGVkcnlFAkTsVBuUqjU6g0mn8ou0pJ0elA%2BlIBndIaMYb9pmwMOFrnEEj5zIQAK4ADyuN2uEOGT0ILzeHwwX0IPzh%2BwB2U6%2BW8ZL0%2B0RQO6vQMkiAA))

If you select a platform in the playground, all errors go away, because this `if`-`else` is already special-cased by the type checker and the code for other platforms is treated as unreachable.

**Comment by robert-bn** [comment]
@erictraut 

>A constant (by definition) can have only one statically-determined value. If it can have different values at runtime based on dynamic behaviors, then it's a variable, not a constant.

That is _one_ definition of constant. Another is an immutable variable (i.e. one whose value doesn't change after definition). 

The _immutable variable_ definition more closely reflects how SCREAMING_SNAKE_CASE variables are commonly used in Python, and actually better reflects what pyright currently deems permissible, given pyright treats them as run-time constants in other contexts, for example permitting:

```py
SOME_CONSTANT = some_impure_function()
```

and 
```py
def some_function(x: object):
    SOME_CONSTANT = x
```

It is also more consistent with the language pyright uses to describe violations of this rule: "reportConstantRedefinition" suggests *redefinition* of constants is forbidden (mutually exclusive code paths that *assign* the same variable do not *reassign* it), not that they must be statically determined.
